iroh-net/src/derp/codec.rs

```diff
@@ -0,0 +1,276 @@
+use anyhow::{bail, ensure};
+use bytes::{Buf, BufMut, Bytes, BytesMut};
+use futures::{Stream, StreamExt};
+use tokio_util::codec::{Decoder, Encoder};
+
+use crate::key::PublicKey;
+
+use super::{FrameType, MAGIC, MAX_FRAME_SIZE, NOT_PREFERRED, PREFERRED};
+
+#[derive(Debug, Default, Clone)]
+pub(crate) struct DerpCodec;
+
+#[derive(Debug)]
+pub(crate) struct Frame {
+    pub(crate) typ: FrameType,
+    pub(crate) content: Bytes,
+}
+
+#[derive(Debug)]
+pub(crate) enum WriteFrame<'a> {
```

Comment:
yeah, planning to do sth like that, and to normalize the slice types as well

Reply:
I tried it out in https://github.com/n0-computer/iroh/pull/1387 . Seems to work just fine. You just have to change a lot of code if you go this way because currently the parsing is all over the place.

Reply:
> You just have to change a lot of code if you go this way because currently the parsing is all over the place.

That's fine, will make it much less error prone and easier to test (and change if needed in the future)